Hybrid model for Light simulation UGR

claudiaalvgar opened the pull request:
Pull Request introducing changes needed for using the new Hybrid model for light propagation for the new larg4 in LArSoft. The idea of this model is to use the Semi-Analytic model for propagating photons inside the active volume and a scaled-down Optical-Library for the LAr volume outside the TPC. This way, we are able to have a complete light simulation by having into account the whole cryostat volume.
For using this model some changes are required in larsim. As a summary, our changes are the following: 
- Our proposal is to remove the functions isScintInActiveVolume and extractActiveVolumes from PDFastSimPAR_module. The reason for that is because, in the new larg4, we need them at IonAndScint stage where we want to decide which propagation model to use, via configuration fhicl file, where we specify the volumes where we want to use one model or the other (Semi-Analytic or OpticalLibrary). These functions are needed in ISCalcCorrelated module, in particular, in the function EFieldAtStep() where we need to distinguish between active and non-active volume in order to set a different electric field value inside and outside the active volume. Outside the TPC active volume we need to be able to set to zero (at 1st approach) the EField by using the function isScintInActiveVolume. 
- We propose to create a new class containing these two functions in larsim/IonizationScintillation/, called ISTPC. The reason to create the new class is in order to avoid repeating the functions in both modules (Correlated and PAR), this way both modules are able to use the functions by inheriting the new class.
- Also,  the object “fActiveVolumes" is common to both modules, so it is also included in the new class. 
- Finally in this proposal we have also included the most updated version of PDFastSimPVS_module, provided by Mu Wei which stores reflected light when using the Optical-Library module.  

Your comments will be very welcome: @pgreen135 @aihimmel

claudiaalvgar committed:
Hybrid Model

claudiaalvgar committed:
Hybrid Model

claudiaalvgar committed:
Hybrid Model

claudiaalvgar committed:
Hybrid Model

claudiaalvgar committed:
Hybrid Model

claudiaalvgar committed:
Hybrid Model

FNALbuild commented:
Can one of the admins verify this patch?

FNALbuild commented:
A new Pull Request was created by  @claudiaalvgar  for develop.

It involves the following packages:

larsim


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//933/console

aihimmel commented:
The principle here seems totally sound to me -- let's definitely move common code to a more generic location so multiple modules can use it. However, using multiple inheritance seems like overkill, and might cause some ART problems down the road. (I guess the L1 managers can comment on that point.) It seems like a singleton class (since "Active Volumes" should be the same throughout the job) would do the job and be more robust.

pgreen135 commented:
The changes from the light simulation side look reasonable to me, I will let others comment though re Alex's concern about the inheritance structure.

@claudiaalvgar I triggered the CIs from sbndcode and this currently fails to build in c7 -- looks like an unused variable, see https://dbweb8.fnal.gov:8443/LarCI/app/build_detail/phase_details?build_id=lar_ci/11536&platform=Linux%203.10.0-1160.21.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof

aihimmel commented:
Maybe I can call in @knoepfel for a more definitive comment on the multiple inheritance implementation?

knoepfel commented:
I commend you, @claudiaalvgar, for recognizing that code should not be duplicated when the same functionality is needed in multiple places.  I do agree with @aihimmel that multiple inheritance is not the best route to go here.  But I will comment more next week on some alternatives that still give the desired improvements without introducing unnecessary entanglements through inheritance.  Stay tuned!

knoepfel commented on larsim/IonizationScintillation/ISCalcCorrelated.cxx:
```diff
@@ -130,13 +138,22 @@ namespace larg4 {
+       if (!fSCE->EnableSimEfieldSCE()) return efield;
+       auto const eFieldOffsets = fSCE->GetEfieldOffsets(edep.MidPoint());
+       return efield * std::hypot(1 + eFieldOffsets.X(), eFieldOffsets.Y(), eFieldOffsets.Z());
+     }
```
This body of code can be shortened to:

```cpp
if (!isScintInActiveVolume(edep.MidPoint())) return 0.;
if (!fSCE->EnableSimEfieldSCE()) return efield;       

auto const eFieldOffsets = fSCE->GetEfieldOffsets(edep.MidPoint());
return efield * std::hypot(1 + eFieldOffsets.X(), eFieldOffsets.Y(), eFieldOffsets.Z());
```

knoepfel commented on larsim/IonizationScintillation/ISTPC.cxx:
```diff
@@ -0,0 +1,67 @@
+  {
+    std::cout << "IonizationAndScintillation/ISTPC Initialize." << std::endl;
+    std::cout << "Initializing the geometry of the detector." << std::endl;
+    geo::GeometryCore const& geom = *(lar::providerFrom<geo::Geometry>());
```
Instead of calling `lar::providerFrom<geo::Geometry>()` inside of this class (thus introducing a dependency on the *art* framework), it would be better to pass it into the class through the constructor.  This would then give you a very clean initialization of the class:

```cpp
ISTPC::ISTPC(geo:GeometryCore const& geom) 
  : fActiveVolumes{extractActiveVolumes(geom)}
{ ... }
```

If possible, also avoid directly logging to `std::cout`, which cannot be easily suppressed by users.  It is better to use the message-facility functionality:

```cpp
mf::LogTrace("IonAndScint") << "IonizationAndScintillation/ISTPC Initialize.\n"
                            << "Initializing the geometry of the detector.";
```

knoepfel commented on larsim/IonizationScintillation/ISTPC.h:
```diff
@@ -0,0 +1,34 @@
+#include "larcorealg/CoreUtils/counter.h"
+#include "larcorealg/CoreUtils/enumerate.h"
+
+#include "messagefacility/MessageLogger/MessageLogger.h"
```
These headers should be moved to the .cxx file, which is where they are required.  Otherwise, all downstream code gains a dependency on them, and that's probably not desired.

knoepfel commented on larsim/IonizationScintillation/ISTPC.h:
```diff
@@ -0,0 +1,34 @@
+    static std::vector<geo::BoxBoundedGeo> extractActiveVolumes(geo::GeometryCore const& geom);
+ 
+      
+  };
```
I would adjust the layout of this class.  The only functionality that is required for downstream code is the `isScintInActiveVolume(...)` function.  To gain maximum flexibility for future development, encapsulate everything else as `private`:

```cpp
class ISTPC {
public:
  explicit ISTPC(geo::GeometryCore const& geom);
  bool isScintInActiveVolume(geo::Point_t const& ScintPoint) const;

private:
  std::vector<geo::BoxBoundedGeo> fActiveVolumes;
  static std::vector<geo::BoxBoundedGeo> extractActiveVolumes(...);
};

knoepfel commented on larsim/IonizationScintillation/ISCalcCorrelated.h:
```diff
@@ -31,7 +32,7 @@
 
 namespace larg4 {
-  class ISCalcCorrelated : public ISCalc {
+  class ISCalcCorrelated : public ISTPC, public ISCalc {
```
See comment below regarding multiple inheritance.

knoepfel commented on larsim/PhotonPropagation/PDFastSimPAR_module.cc:
```diff
@@ -252,7 +251,7 @@ namespace {
 
 namespace phot {
-  class PDFastSimPAR : public art::EDProducer {
+  class PDFastSimPAR : public larg4::ISTPC, public art::EDProducer {
```
In this case, I would not choose multiple inheritance.  Although inheriting from `ISTPC` can shorten the amount of typing, it also complicates the class hierarchy and potentially introduces entanglements that can be difficult to resolve later on.  Instead, I encourage a private data member of this class that serves the same role:

```cpp
class PDFastSimPAR : public art::EDProducer {
public:
  ...
private:
  larg4::ISTPC fISTPC;
  ...
};
```

The `isScintInActiveVolume(...)` calls will then turn into `fISTPC.isScintInActiveVolume(...)`, which is a little more typing, but it avoids the class-layout complications introduced through inheritance.

I would make this change to the `ISCalcCorrelated` class also.

knoepfel commented on larsim/PhotonPropagation/PDFastSimPVS_module.cc:
```diff
@@ -3,7 +3,7 @@
 // Plugin Type: producer
 // File:        PDFastSimPVS_module.cc
 // Description:
-// - acts on sim::SimEnergyDeposit from LArG4Main,
+// - acts on sim::SimEnergyDeposit from LArG4Main, 
 // - simulate (fast, photon visibility service) the OpDet response to optical photons
 // Input: 'sim::SimEnergyDeposit'
 // Output: 'sim::OpDetBacktrackerRecord'
```
I will rely on others to make provide feedback on the changes here in this file.  my only comment is there seem to be some whitespace-only changes, which make it difficult to review the meaningful changes.  This usually happens when you don't have the latest changes from the `develop` branch.

knoepfel reviewed (commented):
Thanks for the PR, Claudia.  My review covers mostly design issues--keeping data members private where possible, avoiding multiple inheritance, etc.  Happy to follow up on anything.

claudiaalvgar committed:
Updates for Hybrid Model

claudiaalvgar commented:
List of modificatios (following @knoepfel suggestions):
-Updated constructor for ISTPC class, receiving geom parameter.
-Moved includes from ISTPC.h to ISTPC.cxx where they are used.
-Removed ISTPC inheritance creating an ISTPC object (ISTPC fISTPC;) in ISCalcCorrelated and in PDFastSimPAR_module to access functions like: fISTPC.isScintInActiveVolume and fISTPC.extractActiveLArVolume.
-Initialization of fISTPC object in ISCalcCorrelated and in PDFastSimPAR_module constructor.
-Change the name of extractActiveVolumes to extractActiveLArVolume.
-Updated EFieldAtStep function in ISCalcCorrelated.cxx.
-Removed unused variable (nion) of PDFastSimPVS_module.

knoepfel reviewed (commented):
Looks good, Claudia.  This will receive L1 approval after L2 approval has been granted.

knoepfel commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
Pull request #68 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//937/console

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11574&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11575&builds= 
 +build

FNALbuild commented:
-argoneut tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5753&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11575&builds=

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5510&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11575&builds=

FNALbuild commented:
-dune tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/11384&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11575&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/6198&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11575&builds=

FNALbuild commented:
-sbnd tests warning with build warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/6526&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11575&builds=

FNALbuild commented:
-icarus tests warning with build warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/1991&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11575&builds=

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11607&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11606&builds= 
 +build

FNALbuild commented:
+sbnd tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/6540&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11606&builds=

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5519&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11606&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5763&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11606&builds=

FNALbuild commented:
-icarus tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/2009&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11606&builds=

FNALbuild commented:
-uboone tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/6208&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11606&builds=

FNALbuild commented:
-dune tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/11406&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11606&builds=

absolution1 commented on larsim/IonizationScintillation/ISCalcCorrelated.cxx:
```diff
@@ -65,6 +66,9 @@ namespace larg4 {
 
+    //calculate recombination survival fraction value inside, otherwise zero
+
+    if(EFieldStep > 0) {
```
Pure style comment but I'm guessing the code enclosed in this if statement should be indented?

absolution1 reviewed (approved):
I had a quick look through and only a small stylistic thing jumped out at me.  
I'll leave @aihimmel to grant L2 approval in this case as he's the expert here.

claudiaalvgar committed:
Indented code in if(EfieldStep>0)

aihimmel commented:
As I mentioned above, I think the idea here is a very good one, and it looks like the implementation concern I raised was addressed, so I'm happy for this to proceed. However, I'm no longer an L2 manager so I cannot formally approve.

FNALbuild commented:
Pull request #68 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//956/console

knoepfel commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11611&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11612&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5523&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11612&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5767&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11612&builds=

lgarren commented:
This PR was approved at the coordination meeting pending the changes which have gone in, so I think we're good once the CI tests finish.  Thanks @aihimmel and @absolution1

FNALbuild commented:
+sbnd tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/6544&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11612&builds=

FNALbuild commented:
-dune tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/11410&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11612&builds=

FNALbuild commented:
-icarus tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/2013&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11612&builds=

FNALbuild commented:
-uboone tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/6212&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/11612&builds=

lgarren commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests.
